File: kumi/sdk/java/src/main/java/io/kumi/Connection.java
package io.kumi;

import java.io.IOException;

/**
 * Connection resolution: determines WebSocket URL and access token
 * from environment variables and connection code.
 */
public final class Connection {
    private Connection() {}

    public static class ConnectionConfig {
        public final String mode;    // "direct" or "relay"
        public final String baseUrl;
        public final String accessToken;

        ConnectionConfig(String mode, String baseUrl, String accessToken) {
            this.mode = mode;
            this.baseUrl = baseUrl;
            this.accessToken = accessToken;
        }

        public String relayEdgeWsUrl() {
            return httpToWs(baseUrl.replaceAll("/+$", "")) + "/ws/edge";
        }
    }

    public static String httpToWs(String url) {
        if (url.startsWith("https://")) return "wss://" + url.substring("https://".length());
        if (url.startsWith("http://")) return "ws://" + url.substring("http://".length());
        return url;
    }

    public static ConnectionConfig resolveConnection(String code, String edgeName)
            throws IOException, InterruptedException {
        String relayUrl = System.getenv("KUMI_RELAY_URL");
        String accessToken = System.getenv("KUMI_ACCESS_TOKEN");
        if (relayUrl != null && !relayUrl.isEmpty() && accessToken != null && !accessToken.isEmpty()) {
            return new ConnectionConfig("relay", relayUrl.replaceAll("/+$", ""), accessToken);
        }

        if (code == null) code = "";

        if (code.startsWith("ws://") || code.startsWith("wss://")) {
            return new ConnectionConfig("direct", code, null);
        }

        if (Auth.isLanCode(code)) {
            String serverUrl = Auth.parseLanCode(code);
            String wsUrl = httpToWs(serverUrl.replaceAll("/+$", "")) + "/ws/edge";
            return new ConnectionConfig("direct", wsUrl, null);
        }

        if (Auth.isRelayToken(code)) {
            Auth.BootstrapResult profile = Auth.bootstrapProfile(code, "edge", edgeName);
            return new ConnectionConfig("relay", profile.relayUrl, profile.accessToken);
        }

        if (code.startsWith("http://") || code.startsWith("https://")) {
            String wsUrl = httpToWs(code.replaceAll("/+$", "")) + "/ws/edge";
            return new ConnectionConfig("direct", wsUrl, null);
        }

        return new ConnectionConfig("direct", "ws://127.0.0.1:8000/ws/edge", null);
    }
}

File: kumi/sdk/kotlin/src/main/kotlin/io/kumi/sdk/Connection.kt
package io.kumi.sdk

data class ConnectionConfig(
    val mode: String,
    val baseUrl: String,
    val accessToken: String,
) {
    fun websocketUrl(): String =
        if (mode == "relay") "${httpToWs(baseUrl.trimEnd('/'))}/ws/edge" else baseUrl
}

fun httpToWs(url: String): String {
    val u = url.trimEnd('/')
    return when {
        u.startsWith("https://") -> "wss://" + u.substring("https://".length)
        u.startsWith("http://") -> "ws://" + u.substring("http://".length)
        else -> u
    }
}

fun resolveConnection(code: String, edgeName: String): ConnectionConfig {
    val relayUrl = kumiEnv("KUMI_RELAY_URL")
    val accessToken = kumiEnv("KUMI_ACCESS_TOKEN")
    if (relayUrl.isNotEmpty() && accessToken.isNotEmpty()) {
        return ConnectionConfig("relay", relayUrl.trimEnd('/'), accessToken)
    }

    if (code.startsWith("ws://") || code.startsWith("wss://")) {
        return ConnectionConfig("direct", code, "")
    }

    if (isLanCode(code)) {
        val serverUrl = parseLanCodeToServerUrl(code)
        val wsUrl = "${httpToWs(serverUrl)}/ws/edge"
        return ConnectionConfig("direct", wsUrl, "")
    }

    if (isRelayToken(code)) {
        val profile = bootstrapProfile(code, "edge", edgeName)
        setKumiEnv("KUMI_RELAY_URL", profile.relayUrl)
        setKumiEnv("KUMI_ACCESS_TOKEN", profile.accessToken)
        return ConnectionConfig("relay", profile.relayUrl, profile.accessToken)
    }

    if (code.startsWith("http://") || code.startsWith("https://")) {
        val base = code.trimEnd('/')
        val wsUrl = "${httpToWs(base)}/ws/edge"
        return ConnectionConfig("direct", wsUrl, "")
    }

    return ConnectionConfig("direct", "ws://127.0.0.1:8000/ws/edge", "")
}
